storage/cleanup.go

```diff
@@ -0,0 +1,159 @@
+	fm                      databaseFlushManager
+	filesetFilesBeforeFn    filesetFilesBeforeFn
+	commitLogFilesBeforeFn  commitLogFilesBeforeFn
+	commitLogFilesForTimeFn commitLogFilesForTimeFn
+	deleteFilesFn           deleteFilesFn
+}
+
+func newCleanupManager(database database, fm databaseFlushManager) databaseCleanupManager {
+	opts := database.Options()
+	blockSize := opts.GetRetentionOptions().GetBlockSize()
+	filePathPrefix := opts.GetCommitLogOptions().GetFilesystemOptions().GetFilePathPrefix()
+	commitLogsDir := fs.CommitLogsDirPath(filePathPrefix)
+
+	return &cleanupManager{
+		database:                database,
+		opts:                    opts,
+		blockSize:               blockSize,
+		filePathPrefix:          filePathPrefix,
+		commitLogsDir:           commitLogsDir,
+		fm:                      fm,
+		filesetFilesBeforeFn:    fs.FilesetFilesBefore,
+		commitLogFilesBeforeFn:  fs.CommitLogFilesBefore,
+		commitLogFilesForTimeFn: fs.CommitLogFilesForTime,
+		deleteFilesFn:           fs.DeleteFiles,
+	}
+}
+
+func (mgr *cleanupManager) Cleanup(t time.Time) error {
+	multiErr := xerrors.NewMultiError()
+	filesetFilesStart := mgr.fm.FlushTimeStart(t)
+	if err := mgr.cleanupFilesetFiles(filesetFilesStart); err != nil {
+		detailedErr := fmt.Errorf("encountered errors when cleaning up fileset files for %v: %v", filesetFilesStart, err)
+		multiErr = multiErr.Add(detailedErr)
+	}
+	commitLogStart, commitLogTimes := mgr.commitLogTimes(t)
+	if err := mgr.cleanupCommitLogs(commitLogStart, commitLogTimes); err != nil {
+		detailedErr := fmt.Errorf("encountered errors when cleaning up commit logs for commitLogStart %v commitLogTimes %v: %v", commitLogStart, commitLogTimes, err)
+		multiErr = multiErr.Add(detailedErr)
+	}
+	return multiErr.FinalError()
+}
+
+func (mgr *cleanupManager) cleanupFilesetFiles(earliestToRetain time.Time) error {
+	multiErr := xerrors.NewMultiError()
+	shards := mgr.database.getOwnedShards()
+	for _, shard := range shards {
+		shardID := shard.ShardNum()
+		expired, err := mgr.filesetFilesBeforeFn(mgr.filePathPrefix, shardID, earliestToRetain)
+		if err != nil {
+			detailedErr := fmt.Errorf("encountered errors when getting fileset files for prefix %s shard %d: %v", mgr.filePathPrefix, shardID, err)
+			multiErr = multiErr.Add(detailedErr)
+		}
+		if err := mgr.deleteFilesFn(expired); err != nil {
+			multiErr = multiErr.Add(err)
+		}
+	}
+
+	return multiErr.FinalError()
+}
+
+// NB(xichen): since each commit log contains data needed for bootstrapping not only
+// its own block size period but also its left and right block neighbors due to past
+// writes and future writes, we need to shift flush time range by block size as the
+// time range for commit log files we need to check.
+func (mgr *cleanupManager) commitLogTimeRange(t time.Time) (time.Time, time.Time) {
+	flushStart, flushEnd := mgr.fm.FlushTimeStart(t), mgr.fm.FlushTimeEnd(t)
+	return flushStart.Add(-mgr.blockSize), flushEnd.Add(-mgr.blockSize)
```

Comment:
If we need to do right neighbour shouldn't the second return arg be `flushEnd.Add(m.blockSize)`?